serde-reflection/src/de.rs

```diff
@@ -391,55 +396,147 @@ impl<'de, 'a> de::Deserializer<'de> for Deserializer<'de, 'a> {
 
     // Assumption: The first variant(s) should be "base cases", i.e. not cause infinite recursion
     // while constructing sample values.
+    #[allow(clippy::map_entry)] // false positive https://github.com/rust-lang/rust-clippy/issues/9470
     fn deserialize_enum<V>(
         self,
-        name: &'static str,
+        enum_name: &'static str,
         variants: &'static [&'static str],
         visitor: V,
     ) -> Result<V::Value>
     where
         V: Visitor<'de>,
     {
-        self.format.unify(Format::TypeName(name.into()))?;
+        let enum_type_id = typeid::of::<V::Value>();
+        self.format.unify(Format::TypeName(enum_name.into()))?;
         // Pre-update the registry.
         self.tracer
             .registry
-            .entry(name.to_string())
+            .entry(enum_name.to_string())
             .unify(ContainerFormat::Enum(BTreeMap::new()))?;
-        let known_variants = match self.tracer.registry.get_mut(name) {
+        let known_variants = match self.tracer.registry.get_mut(enum_name) {
             Some(ContainerFormat::Enum(x)) => x,
             _ => unreachable!(),
         };
-        // If we have found all the variants OR if the enum is marked as
-        // incomplete already, pick the first index.
-        let index = if known_variants.len() == variants.len()
-            || self.tracer.incomplete_enums.contains(name)
-        {
-            0
-        } else {
-            let mut index = known_variants.len() as u32;
-            // Scan the range 0..=known_variants.len() downwards to find the next
-            // variant index to explore.
-            while known_variants.contains_key(&index) {
-                index -= 1;
+
+        // If the enum is marked as incomplete, just visit the first index
+        // because we presume it avoids recursion.
+        if self.tracer.incomplete_enums.contains_key(enum_name) {
```

Comment:
Perhaps also match `variants.len() == 0` while we're at it (to avoid an underflow below)

Reply:
For performance reasons (in project where there are many enums), it would be nice to have a criteria to quickly skip the analysis when it was completed previously.

I wonder if we could structure `self.tracer.discriminants` with nested maps and write a stop condition like:
```
self.tracer.discriminants.get(&enum_type_id).or_default().len() == variants.len() && known_variants.range(provisional_min..).is_empty()
```

Reply:
Added a check for `variants.is_empty()`. Before this PR, this would have panicked. After this PR it will return a NotSupported error.